File: docs/closures/program_correct.cpp
// g++ -std=c++20 -o out_cpp program_correct.cpp && ./out_cpp

#include <functional>
#include <iostream>
#include <memory>

struct Counter {
  std::function<int(int)> inc;
  std::function<void()> reset;
};

Counter makeCounter(int initial) {
  auto counter = std::make_shared<int>(initial);

  auto inc = [=](int by) mutable -> int {
    *counter += by;
    return *counter;
  };

  auto reset = [=]() mutable { *counter = initial; };

  return (Counter){.inc = inc, .reset = reset};
}

int main() {
  auto counter = makeCounter(0);
  std::cout << counter.inc(42) << std::endl; // 42
  std::cout << counter.inc(42) << std::endl; // 84
  counter.reset();
  std::cout << counter.inc(1) << std::endl; // 1
}

File: docs/closures/program_correct.rs
// rustc -o out_rs program_correct.rs && ./out_rs

use std::cell::RefCell;
use std::rc::Rc;

struct Counter {
  inc: Box<dyn FnMut(i32) -> i32>,
  reset: Box<dyn FnMut()>,
}

fn make_counter(initial: i32) -> Counter {
  let counter = Rc::new(RefCell::new(initial));

  let inc = {
    let counter = Rc::clone(&counter);
    Box::new(move |by| {
      let mut counter = counter.borrow_mut();
      *counter += by;
      *counter
    })
  };

  let reset = {
    let counter = Rc::clone(&counter);
    Box::new(move || {
      let mut counter = counter.borrow_mut();
      *counter = initial;
    })
  };

  Counter { inc, reset }
}

fn main() {
  let mut counter = make_counter(0);
  println!("{}", (counter.inc)(42)); // 42
  println!("{}", (counter.inc)(42)); // 84
  (counter.reset)();
  println!("{}", (counter.inc)(1)); // 1
}